miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultLibraryService.java

```diff
@@ -388,6 +395,53 @@ public void deleteQc(Library library, Long qcId) throws IOException {
   }
 
   /**
+   * Turns indices into strings for easier comparison and changelog message concatentation.
+   * 
+   * @param indices
+   * @return
+   */
+  private Set<String> stringifyIndices(List<Index> indices) {
+    Set<String> original = new HashSet<>();
+    for (Index index : indices) {
+      if (index != null) {
+        original.add(index.getFamily().getName() + " - " + index.getLabel());
+      }
+    }
+    return original;
+  }
+
+  /**
+   * Create a changelog if the indices have changed.
+   * 
+   * @param originalIndices
+   * @param updatedIndices
+   * @param target
+   * @throws IOException
+   */
+  private void makeChangeLogForIndices(List<Index> originalIndices, List<Index> updatedIndices, Library target) throws IOException {
+    
+    Set<String> added = new TreeSet<>(stringifyIndices(updatedIndices));
+    added.removeAll(originalIndices);
```

Comment:
You're removing `Index` objects from a set containing strings. You probably want to do:
```
Set<String> original  = stringifyIndices(originalIndices);
Set<String> updated  = stringifyIndices(updatedIndices);
Set<String> added = new TreeSet<>(updated);
added.removeAll(original);
Set<String> removed = new TreeSet<>(original);
removed.removeAll(updated);
```